modules/core/src/main/java/org/apache/ignite/internal/binary/streams/BinaryMemoryAllocator.java

```diff
@@ -16,41 +16,258 @@
+         *
+         * @return {@code true} if acquired {@code false} otherwise.
+         */
+        @Override public boolean isAcquired() {
+            BinaryMemoryAllocatorChunk holder = holders.get();
+
+            return holder != null && holder.isAcquired();
+        }
+
+        /**
+         * Memory allocator chunk.
+         */
+        private static class Chunk implements BinaryMemoryAllocatorChunk {
+            /** Data array */
+            private byte[] data;
+
+            /** Max message size detected between checks. */
+            private int maxMsgSize;
+
+            /** Last time array size is checked. */
+            private long lastCheckNanos = System.nanoTime();
+
+            /** Whether the holder is acquired or not. */
+            private boolean acquired;
+
+            /** {@inheritDoc} */
+            @Override public byte[] allocate(int size) {
+                if (acquired)
+                    return new byte[size];
+
+                acquired = true;
+
+                if (data == null || size > data.length)
+                    data = new byte[size];
+
+                return data;
+            }
+
+            /** {@inheritDoc} */
+            @Override public byte[] reallocate(byte[] data, int size) {
+                byte[] newData = new byte[size];
+
+                if (this.data == data)
+                    this.data = newData;
+
+                System.arraycopy(data, 0, newData, 0, data.length);
+
+                return newData;
+            }
+
+            /** {@inheritDoc} */
+            @Override public void release(byte[] data, int maxMsgSize) {
+                if (this.data != data)
+                    return;
+
+                if (maxMsgSize > this.maxMsgSize)
+                    this.maxMsgSize = maxMsgSize;
+
+                acquired = false;
 
-    public BinaryMemoryAllocatorChunk chunk() {
-        BinaryMemoryAllocatorChunk holder = holders.get();
+                long nowNanos = System.nanoTime();
 
-        if (holder == null)
-            holders.set(holder = new BinaryMemoryAllocatorChunk());
+                if (U.nanosToMillis(nowNanos - lastCheckNanos) >= CHECK_FREQ) {
+                    int halfSize = data.length >> 1;
 
-        return holder;
+                    if (this.maxMsgSize < halfSize)
+                        this.data = new byte[halfSize];
+
+                    lastCheckNanos = nowNanos;
+                }
+            }
+
+            /** {@inheritDoc} */
+            @Override public boolean isAcquired() {
+                return acquired;
+            }
+        }
     }
 
-    /**
-     * Checks whether a thread-local array is acquired or not.
-     * The function is used by Unit tests.
-     *
-     * @return {@code true} if acquired {@code false} otherwise.
-     */
-    public boolean isAcquired() {
-        BinaryMemoryAllocatorChunk holder = holders.get();
+    /** */
+    private static class PooledAllocator extends BinaryMemoryAllocator {
```

Comment:
Don't we have ready-to-use classes like this already?

Reply:
Nope